File: 2019/2.py
prog = [1, 0, 0, 3, 1, 1, 2, 3, 1, 3, 4, 3, 1, 5, 0, 3, 2, 13, 1, 19, 1, 19, 10, 23, 2, 10, 23, 27, 1, 27, 6, 31, 1, 13, 31, 35, 1, 13, 35, 39, 1, 39, 10, 43, 2, 43, 13, 47, 1, 47, 9, 51, 2, 51, 13, 55, 1, 5, 55, 59, 2, 59, 9, 63, 1, 13, 63, 67,
        2, 13, 67, 71, 1, 71, 5, 75, 2, 75, 13, 79, 1, 79, 6, 83, 1, 83, 5, 87, 2, 87, 6, 91, 1, 5, 91, 95, 1, 95, 13, 99, 2, 99, 6, 103, 1, 5, 103, 107, 1, 107, 9, 111, 2, 6, 111, 115, 1, 5, 115, 119, 1, 119, 2, 123, 1, 6, 123, 0, 99, 2, 14, 0, 0]


def run(mem):
    i = 0
    while (op := mem[i]) != 99:

        p1 = mem[i+1]
        p2 = mem[i+2]
        p3 = mem[i+3]
        if op == 1:
            mem[p3] = mem[p1] + mem[p2]

        if op == 2:
            mem[p3] = mem[p1] * mem[p2]

        i += 4

    return mem[0]


# part 1
mem1 = prog.copy()
mem1[1] = 12
mem1[2] = 2
print("part1:", run(mem1))


# part 2
for noun in range(0, 100):
    for verb in range(0, 100):
        mem2 = prog.copy()
        mem2[1] = noun
        mem2[2] = verb
        if run(mem2) == 19690720:
            print("part2: noun= {} verb= {} res= {}".format(
                noun, verb, 100 * noun + verb))

File: 2019/src/bin/2.rs
fn main() {
    let input: Vec<usize> = std::fs::read_to_string("./input/input-2.txt")
        .unwrap()
        .split(',')
        .filter_map(|l| l.parse().ok())
        .collect();

    let res = run_program(input.clone(), 12, 2);
    println!("part1: {}", res);

    for noun in 0..=99 {
        for verb in 0..=99 {
            let res = run_program(input.clone(), noun, verb);
            if res == 19690720 {
                // println!("noun= {} verb= {} => {}", noun, verb, 100 * noun + verb);
                println!("part2: {}", 100 * noun + verb);

                break;
            }
        }
    }
}

fn run_program(mut input: Vec<usize>, noun: usize, verb: usize) -> usize {
    input[1] = noun;
    input[2] = verb;
    let mut i: usize = 0;
    loop {
        match input[i] {
            1 => {
                let add1 = input[i + 1];
                let add2 = input[i + 2];
                let res = input[i + 3];
                input[res] = input[add1] + input[add2];
            }
            2 => {
                let mul1 = input[i + 1];
                let mul2 = input[i + 2];
                let res = input[i + 3];
                input[res] = input[mul1] * input[mul2];
            }
            99 => {
                break;
            }
            _ => panic!("unknown opcode"),
        }
        i += 4;
    }

    input[0]
}
